File: get_picata_catalog.php
<?php

	include "main_library.php";
	$logs=null;
	$id_user=null;
	init_logs();

	$id = '3991533';
	$lien="https://www.picata.fr/tarifs/Tarif_ean.csv";
	//$lien="picata_catalog.txt";
	$filename = "picata_catalog.xml";
	
	$to_remove = array('"', '=');
	$to_remplace = array(" ", "?");
	
	$output = file_get_contents($lien);
	$output = iconv('ISO-8859-15', 'UTF-8//TRANSLIT', $output);
	$rows = explode("\n",$output);
	$first_line = array_shift($rows);
	$first_line = substr($first_line,0, strlen ($first_line) - 1); 
	$criterias = explode(";",str_replace($to_remplace,"_",$first_line));
	// $criterias = array("Ean","Reference","Categorie","Constructeur","Libelle","PxHT","StockDispoAchard","EncdeAchard");

	$xml_string = "<?xml version=\"1.0\"?>\n<customer_products>\n";
	foreach($rows as $row) {
		$row = substr($row,0, strlen ($row) - 1);
		$datas = explode(";",$row);
		// prevent malformed product
		if (count($datas) == count($criterias)){
			$xml_string = $xml_string."<customer_product>\n";
			for ($i = 0; $i < (count($datas)); $i++) {
				if ($i == 0)
					$datas[$i] = str_replace($to_remove, "", $datas[$i]);
				$datas[$i] = str_replace("&", "+", $datas[$i]);
				$xml_string = $xml_string."<$criterias[$i]>$datas[$i]</$criterias[$i]>";
			}
			$xml_string = $xml_string."\n</customer_product>\n";
		}
	}
	$xml_string = $xml_string."</customer_products>";
	file_put_contents($filename,$xml_string);

?>

File: get_picata_catalog.py
import myLib
import log

from lxml import etree
import requests
import re
import os
import io

FILENAME = "picata_catalog.xml"
BLACKLIST_FILENAME = "categories_blacklisted.txt"
url = "https://www.picata.fr/tarifs/Tarif_ean.csv"

PROVIDER_TAG = "PIC"

TAGS = ["barcode","reference","product_category","brand","name","provider_price","stock_dispo","stock_cmd"]

blacklist_items = []
print("--BLACKLIST--")
with open(BLACKLIST_FILENAME, "r") as blacklist_file:
    for blacklist_item in blacklist_file:
        blacklist_items.append(blacklist_item.strip())
        print(blacklist_item)
print("--END_BLKLS--")



r = requests.get(url)
lines = r.content.split("\r\n")
lines.pop(0) # On retire la premier ligne, l'organisation du CSV doit etre parametre via la var TAGS ci dessus
print len(lines)
with io.open(FILENAME, "w", encoding="utf8") as f2:
    f2.write(u"<?xml version=\"1.0\"?>\n<customer_products>\n")
    for line in lines:
        items = line.strip().split(";")
        if(len(TAGS) != len(items)):
            log.warning("CSV line is not correctly formatted :\n{}".format(line))
            continue
        string = ""
        for i in range(len(TAGS)):
            # Premier traitement : Si l'item contient un signe @
            value=items[i].replace('&','&amp;')
            # Deuxieme traitement : Si l'item contient deux signes "" (le premier servant a echapper le deuxieme)
            if ( re.match("^\".*\"{2}.*\"$", value)):
                value = re.sub("\"{2}","\'",value)
                value = value.replace("\"","").replace("\'","\"")
            # Troisieme traitement : On ajoute le tag fournisseur sur la ref
            if (TAGS[i] == "reference"):
                value = PROVIDER_TAG+value
            # Quatrieme traitement : pour l'EAN, on supprime l'encapsulage ="XXXXXXXXXXX"
            if (TAGS[i] == "barcode"):
                if ( re.match("^=\"\d*\"$", value)):
                    value = re.sub("={0,1}\"","",value)
                else:
                    log.warning("product with ref {} don't respect EAN expected format : {}".format(items[1], value))
            # Cinquieme traitement : si la categorie fait partie de celles blackliste
            if (TAGS[i] == "product_category"):
                if value in blacklist_items:
                    log.warning("product with ref {} skipped for being in the blacklist".format(items[1]))
                    continue
                
            string = string+"<"+TAGS[i]+">"+value+"</"+TAGS[i]+">"
        f2.write(u"<customer_product>\n")
        f2.write(string.decode('ISO-8859-15'))
        f2.write(u"\n</customer_product>\n")
    f2.write(u"</customer_products>")